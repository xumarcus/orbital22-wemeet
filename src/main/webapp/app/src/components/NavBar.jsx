import MenuIcon from '@mui/icons-material/Menu';
import AppBar from '@mui/material/AppBar';
import Avatar from '@mui/material/Avatar';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Container from '@mui/material/Container';
import IconButton from '@mui/material/IconButton';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import Toolbar from '@mui/material/Toolbar';
import Tooltip from '@mui/material/Tooltip';
import Typography from '@mui/material/Typography';
import * as React from 'react';
import {useContext, useState} from 'react';
import logo from './logo_banner.png';
import {Link, useNavigate} from 'react-router-dom';
import LogInModal from './LogInModal';
import ForgetPasswordModel from './ForgetPasswordModal';
import SignUpModal from './SignUpModal';
import AppContext from '../core/app-context';
import ajax from '../core/util';

const pages = [
  ['Features', 'features'],
  ['Guide', 'guide'],
  ['About Us', 'about'],
];

const settings = [
  ['Dashboard', 'dashboard'],
  ['Profile', 'profile'],
  ['Log out', 'logout'],
];

const NavBar = () => {
  const appContext = useContext(AppContext);

  const [anchorElNav, setAnchorElNav] = useState(null);
  const [anchorElUser, setAnchorElUser] = useState(null);
  const [ModalVisible, setModalVisible] = useState('');

  const navigate = useNavigate();

  const handleOpenNavMenu = (event) => {
    setAnchorElNav(event.currentTarget);
  };
  const handleOpenUserMenu = (event) => {
    setAnchorElUser(event.currentTarget);
  };

  const handleCloseNavMenu = (newPage) => {
    setAnchorElNav(null);
    navigate(`/${newPage}`);
  };

  const handleCloseUserMenu = async (newPage) => {
    setAnchorElUser(null);
    if (newPage === 'logout') {
      await ajax('POST')('/logout');
      appContext.setValues({...appContext.values, user: null});
    } else {
      navigate(`/${newPage}`);
    }
  };

  const handleStartNowClick = () => {
    console.log('Start Now');
    setModalVisible('signin');
  };

  return (
      <>
        <LogInModal visible={ModalVisible} setVisible={setModalVisible}/>
        <ForgetPasswordModel
            visible={ModalVisible}
            setVisible={setModalVisible}
        />
        <SignUpModal visible={ModalVisible} setVisible={setModalVisible}/>
        <AppBar position="static" sx={{bgcolor: 'white', boxShadow: '0'}}>
          <Container maxWidth="xl">
            <Toolbar disableGutters>
              <Link to={'/'}>
                <Box
                    component="img"
                    sx={{
                      display: {xs: 'none', md: 'flex'},
                      my: 3,
                      mx: 2,
                      maxHeight: {xs: 320, md: 250},
                      maxWidth: {xs: 350, md: 300},
                    }}
                    alt="WeMeet"
                    src={logo}
                />
              </Link>
              <Box sx={{flexGrow: 1, display: {xs: 'flex', md: 'none'}}}>
                <IconButton
                    size="large"
                    aria-label="account of current user"
                    aria-controls="menu-appbar"
                    aria-haspopup="true"
                    onClick={handleOpenNavMenu}
                    color="secondary"
                >
                  <MenuIcon/>
                </IconButton>
                <Menu
                    id="menu-appbar"
                    anchorEl={anchorElNav}
                    anchorOrigin={{
                      vertical: 'bottom',
                      horizontal: 'left',
                    }}
                    keepMounted
                    transformOrigin={{
                      vertical: 'top',
                      horizontal: 'left',
                    }}
                    open={Boolean(anchorElNav)}
                    onClose={handleCloseNavMenu}
                    sx={{
                      display: {xs: 'block', md: 'none'},
                    }}
                >
                  {pages.map((page) => (
                      <MenuItem
                          key={page}
                          onClick={handleCloseNavMenu.bind(this, page[1])}
                      >
                        <Typography textAlign="center">{page[0]}</Typography>
                      </MenuItem>
                  ))}
                </Menu>
              </Box>
              <Box
                  sx={{flexGrow: 1, display: {xs: 'none', md: 'flex'}, pr: 2}}
                  justifyContent="flex-end"
              >
                {pages.map((page) => (
                    <Button
                        key={page}
                        onClick={handleCloseNavMenu.bind(this, page[1])}
                        sx={{
                          my: 2,
                          color: 'black',
                          fontWeight: 'bold',
                          display: 'block',
                        }}
                    >
                      {page[1]}
                    </Button>
                ))}
              </Box>

              <Box sx={{flexGrow: 0}}>
                <Tooltip title="Open settings">
                  <IconButton onClick={handleOpenUserMenu} sx={{p: 0}}>
                    {/* FIXME */}
                    <Avatar alt={appContext.values.user?.email ?? 'Anonymous'}
                            src={'404'}/>
                  </IconButton>
                </Tooltip>
                <Menu
                    sx={{mt: '45px'}}
                    id="menu-appbar"
                    anchorEl={anchorElUser}
                    anchorOrigin={{
                      vertical: 'top',
                      horizontal: 'right',
                    }}
                    keepMounted
                    transformOrigin={{
                      vertical: 'top',
                      horizontal: 'right',
                    }}
                    open={Boolean(anchorElUser)}
                    onClose={handleCloseUserMenu}
                >
                  {settings.map((setting) => (
                      <MenuItem key={setting[0]}
                                onClick={handleCloseUserMenu.bind(this,
                                    setting[1])}>
                        <Typography textAlign="center">{setting[0]}</Typography>
                      </MenuItem>
                  ))}
                </Menu>
              </Box>
              <Box sx={{pl: 2}}>
                <Button
                    variant="contained"
                    size="large"
                    onClick={handleStartNowClick}
                    color="success"
                >
                  Start Now
                </Button>
              </Box>
            </Toolbar>
          </Container>
        </AppBar>
      </>
  );
};
export default NavBar;
