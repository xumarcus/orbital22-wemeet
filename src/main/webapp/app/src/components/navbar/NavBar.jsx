import MenuIcon from '@mui/icons-material/Menu'
import AppBar from '@mui/material/AppBar'
import Avatar from '@mui/material/Avatar'
import Box from '@mui/material/Box'
import Button from '@mui/material/Button'
import Container from '@mui/material/Container'
import IconButton from '@mui/material/IconButton'
import Menu from '@mui/material/Menu'
import MenuItem from '@mui/material/MenuItem'
import Toolbar from '@mui/material/Toolbar'
import Tooltip from '@mui/material/Tooltip'
import Typography from '@mui/material/Typography'
import * as React from 'react'
import { useContext, useState } from 'react'
import { Link, useNavigate } from 'react-router-dom'
import LogInModal from './authentication/LogInModal'
import SignUpModal from './authentication/SignUpModal'
import AppContext, { defaultAppContext } from '../../core/AppContext'
import ajax from '../../core/ajax'
import WeMeetLogo from './WeMeetLogo'
import MaterialLink from '@mui/material/Link'

/*
const pages = [
  ['Features', 'features'],
  ['Guide', 'guide'],
  ['About Us', 'about']
]

 */

const settings = [
  ['Dashboard', 'dashboard'],
  // TODO ['Profile', 'profile'],
  ['Log out', 'logout']
]

const NavBar = () => {
  const { context, setContext } = useContext(AppContext)

  const [anchorElNav, setAnchorElNav] = useState(null)
  const [anchorElUser, setAnchorElUser] = useState(null)
  const [ModalVisible, setModalVisible] = useState('')

  const navigate = useNavigate()

  const handleOpenNavMenu = (event) => {
    setAnchorElNav(event.currentTarget)
  }
  const handleOpenUserMenu = (event) => {
    setAnchorElUser(event.currentTarget)
  }

  const handleCloseNavMenu = (newPage) => {
    setAnchorElNav(null)
    navigate(`/${newPage}`)
  }

  const handleCloseUserMenu = async (newPage) => {
    setAnchorElUser(null)
    if (newPage === 'logout') {
      try {
        await ajax('POST')('/logout')
      } catch (e) {
        // do nothing
      } finally {
        setContext(defaultAppContext)
      }
    } else {
      navigate(`/${newPage}`)
    }
  }

  const handleStartNowClick = () => {
    setModalVisible('signin')
  }

  return (
    <>
      <LogInModal visible={ModalVisible} setVisible={setModalVisible} />
      {/* TODO add forget password support
      <ForgetPasswordModel
        visible={ModalVisible}
        setVisible={setModalVisible}
      />
      */}
      <SignUpModal visible={ModalVisible} setVisible={setModalVisible} />
      <AppBar position='static' sx={{ bgcolor: 'white', boxShadow: '0' }}>
        <Container maxWidth='xl'>
          <Toolbar disableGutters>
            <Link to='/'>
              <WeMeetLogo />
            </Link>
            <Box sx={{ flexGrow: 1, display: { xs: 'flex', md: 'none' } }}>
              <IconButton
                size='large'
                aria-label='account of current user'
                aria-controls='menu-appbar'
                aria-haspopup='true'
                onClick={handleOpenNavMenu}
                color='secondary'
              >
                <MenuIcon />
              </IconButton>
              <Menu
                id='menu-appbar'
                anchorEl={anchorElNav}
                anchorOrigin={{
                  vertical: 'bottom',
                  horizontal: 'left'
                }}
                keepMounted
                transformOrigin={{
                  vertical: 'top',
                  horizontal: 'left'
                }}
                open={Boolean(anchorElNav)}
                onClose={handleCloseNavMenu}
                sx={{
                  display: { xs: 'block', md: 'none' }
                }}
              >
                {/* FIXME comment out
                pages.map((page) => (
                  <MenuItem
                    key={page[1]}
                    onClick={handleCloseNavMenu.bind(this, page[1])}
                  >
                    <Typography textAlign='center'>{page[0]}</Typography>
                  </MenuItem>
                )) */}
                <MenuItem>
                  <MaterialLink href='/static/docs/index.html'>Guide</MaterialLink>
                </MenuItem>
              </Menu>
            </Box>
            <Box
              sx={{ flexGrow: 1, display: { xs: 'none', md: 'flex' }, pr: 2 }}
              justifyContent='flex-end'
            >
              {/* FIXME comment out
              pages.map((page) => (
                <Button
                  key={page}
                  onClick={handleCloseNavMenu.bind(this, page[1])}
                  sx={{
                    my: 2,
                    color: 'black',
                    fontWeight: 'bold',
                    display: 'block'
                  }}
                >
                  {page[1]}
                </Button>
              )) */}
              <Button
                sx={{
                  my: 2,
                  color: 'black',
                  fontWeight: 'bold',
                  display: 'block'
                }}
                href='/static/docs/index.html'
              >Guide
              </Button>
            </Box>

            <Box sx={{ flexGrow: 0 }}>
              <Tooltip title='Open settings'>
                <IconButton onClick={handleOpenUserMenu} sx={{ p: 0 }}>
                  {/* TODO add avatar support */}
                  <Avatar
                    alt={context.user?._links.self.href ?? 'Anonymous'}
                    src='/avatar-does-not-exist'
                  />
                </IconButton>
              </Tooltip>
              <Menu
                sx={{ mt: '45px' }}
                id='menu-appbar'
                anchorEl={anchorElUser}
                anchorOrigin={{
                  vertical: 'top',
                  horizontal: 'right'
                }}
                keepMounted
                transformOrigin={{
                  vertical: 'top',
                  horizontal: 'right'
                }}
                open={Boolean(anchorElUser)}
                onClose={() => setAnchorElUser(null)}
              >
                {settings.map((setting) => (
                  <MenuItem
                    key={setting[0]}
                    onClick={handleCloseUserMenu.bind(this,
                      setting[1])}
                  >
                    <Typography textAlign='center'>{setting[0]}</Typography>
                  </MenuItem>
                ))}
              </Menu>
            </Box>
            <Box sx={{ pl: 2 }}>
              <Button
                variant='contained'
                size='large'
                onClick={handleStartNowClick}
                color='success'
              >
                Start Now
              </Button>
            </Box>
          </Toolbar>
        </Container>
      </AppBar>
    </>
  )
}
export default NavBar
