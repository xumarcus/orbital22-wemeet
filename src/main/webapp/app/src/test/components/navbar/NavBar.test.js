import React from 'react'
import { render } from '@testing-library/react'
import NavBar from '../../../components/navbar/NavBar'
import { BrowserRouter } from 'react-router-dom'

/*
  There is not much point of unit testing when there is not much
  custom UI logic or client side validation. End-to-end tests are
  far more useful, but (...) asks us to have these tests anyway...
 */

describe('NavBar', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<BrowserRouter><NavBar /></BrowserRouter>)
    expect(asFragment()).toMatchSnapshot()
  })
})
