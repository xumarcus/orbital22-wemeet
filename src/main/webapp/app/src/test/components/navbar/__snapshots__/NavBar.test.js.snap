// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should render correctly 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-uv47wz-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
      >
        <a
          href="/"
        >
          <img
            alt="WeMeet"
            class="MuiBox-root css-1pgf9wo"
            src="logo_banner.png"
          />
        </a>
        <div
          class="MuiBox-root css-1lvtzne"
        >
          <button
            aria-controls="menu-appbar"
            aria-haspopup="true"
            aria-label="account of current user"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeLarge css-1f88mc5-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MenuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-2uchni"
        >
          <button
            aria-label="Open settings"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-av538e-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
            >
              <img
                alt="Anonymous"
                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                src="/avatar-does-not-exist"
              />
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-n085mf"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root  css-mkazzp-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Start Now
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
